retrievalmarket/impl/requestvalidation/revalidator.go

```diff
@@ -178,12 +178,21 @@ func (pr *ProviderRevalidator) processPayment(dealID rm.ProviderDealIdentifier,
 
 	// resume deal
 	_ = pr.env.SendEvent(dealID, rm.ProviderEventPaymentReceived, received)
+
 	if deal.Status == rm.DealStatusFundsNeededLastPayment {
 		return &rm.DealResponse{
 			ID:     deal.ID,
 			Status: rm.DealStatusCompleted,
 		}, nil
 	}
+
+	// This is hacky. But, if we don't do this, the data transfer module will resume data transfer
+	// even though we haven't finished reading the Unsealed data into the local Block-store since data transfer
+	// treats a "nil" error from here as a resumption signal.
+	if deal.Status == rm.DealStatusUnsealing || deal.Status == rm.DealStatusFundsNeededUnseal {
+		return nil, datatransfer.ErrNoOp
+	}
+
 	return nil, nil
```

Comment:
IIUC in the worst case the provider may unpause too early, accidentally sending a bit of extra data to the client, but that's fine, this code _does_ guarantee that unpause only happens after unsealing has completed which is the important part 👍